zerver/lib/user_topics.py

```diff
@@ -95,39 +96,46 @@ def set_topic_mutes(
         recipient_id = stream.recipient_id
         assert recipient_id is not None
 
-        add_topic_mute(
+        set_user_topic_visibility_policy(
             user_profile=user_profile,
             stream_id=stream.id,
             recipient_id=recipient_id,
             topic_name=topic_name,
-            date_muted=date_muted,
+            visibility_policy=UserTopic.MUTED,
+            date_added=date_muted,
         )
 
 
-def add_topic_mute(
+@transaction.atomic
+def set_user_topic_visibility_policy(
     user_profile: UserProfile,
     stream_id: int,
-    recipient_id: int,
     topic_name: str,
-    date_muted: Optional[datetime.datetime] = None,
+    *,
+    visibility_policy: int,
+    recipient_id: int,
+    date_added: Optional[datetime.datetime] = None,
     ignore_duplicate: bool = False,
 ) -> None:
-    if date_muted is None:
-        date_muted = timezone_now()
-    UserTopic.objects.bulk_create(
-        [
-            UserTopic(
-                user_profile=user_profile,
-                stream_id=stream_id,
-                recipient_id=recipient_id,
-                topic_name=topic_name,
-                last_updated=date_muted,
-                visibility_policy=UserTopic.MUTED,
-            ),
-        ],
-        ignore_conflicts=ignore_duplicate,
+    if date_added is None:
+        date_added = timezone_now()
+
+    (row, created) = UserTopic.objects.get_or_create(
+        user_profile=user_profile,
+        stream_id=stream_id,
+        topic_name=topic_name,
+        recipient_id=recipient_id,
+        defaults={"last_updated": date_added, "visibility_policy": visibility_policy},
     )
 
+    if not created:
+        if row.visibility_policy == visibility_policy and not ignore_duplicate:
+            raise IntegrityError  # nocoverage
+        else:
```

Comment:
Hmm, this pattern seems off to me. Why should we need to raise an `IntegrityError` ourselves? Isn't that the database's work?

Reply:
Why not `raise JsonableError(_("Topic is not muted"))` here, instead of `raise IntegrityError`? That seems more natural to me, although it may not solve the coverage issue.

Reply:
Maybe its best to not raise anything (to avoid over-complicating things inside a transaction)?

```python
        if not created:
            if row.visibility_policy == visibility_policy:
                if ignore_duplicate:
                    return

            row.visibility_policy = visibility_policy
            row.last_updated = date_added
            row.save(update_fields=["visibility_policy", "last_updated"])
```
Could you check if that works?

Reply:
Ah, actually the `no covereage` error was due to not using **case insensitive** lookup for topic_name in 'get_or_create'
Updating it, fixes the issue.

```diff
get_or_create(
.
-`topic_name=topic_name`
+`topic_name__iexact=topic_name
.
```

Now, addressing the other concerns:
1. If we use the following code, [this piece of test](https://github.com/zulip/zulip/pull/24236/files#diff-dca29140ac4182499e88c086202b897cca41c670b6163288b76f582569528402R126) will fail.
```
        if not created:
            if row.visibility_policy == visibility_policy:
                if ignore_duplicate:
                    return

            row.visibility_policy = visibility_policy
            row.last_updated = date_added
            row.save(update_fields=["visibility_policy", "last_updated"])
```

Reply:
The link to the test seems to be broken. Could you paste the test code (which fails) here?

Reply:
In `zerver/tests/test_user_topics.py` [Line-123](https://github.com/zulip/zulip/blob/main/zerver/tests/test_user_topics.py#L123)
```python
        assert stream.recipient is not None
        result = self.api_patch(user, url, data)
        
        # Now check that error is raised when attempted to mute an already
        # muted topic. This should be case-insensitive.
        data["topic"] = "VERONA3"
        result = self.api_patch(user, url, data)
        self.assert_json_error(result, "Topic already muted") # this assertion fails

```
DB should raise IntegrityError -- as we removed one of the duplicate check in the previous PR #24354

Reply:
Ah, that makes sense :)

Reply:
My concern was that it doesn't seem ergonomic to raise an `IntegrityError` like this artificially. Why not return a usual API error instead?
```diff
diff --git a/zerver/lib/user_topics.py b/zerver/lib/user_topics.py
index fb03bf42fe..febe2cd2c6 100644
--- a/zerver/lib/user_topics.py
+++ b/zerver/lib/user_topics.py
@@ -1,12 +1,14 @@
 import datetime
 from typing import Callable, List, Optional, Tuple, TypedDict

-from django.db import IntegrityError, transaction
+from django.db import transaction
 from django.db.models.query import QuerySet
 from django.utils.timezone import now as timezone_now
+from django.utils.translation import gettext as _
 from sqlalchemy.sql import ClauseElement, and_, column, not_, or_
 from sqlalchemy.types import Integer

+from zerver.lib.exceptions import JsonableError
 from zerver.lib.timestamp import datetime_to_timestamp
 from zerver.lib.topic import topic_match_sa
 from zerver.lib.types import UserTopicDict
@@ -142,7 +144,10 @@ def set_user_topic_visibility_policy(

         if not created:
             if row.visibility_policy == visibility_policy and not ignore_duplicate:
-                raise IntegrityError
+                # Duplicate request case (topic-visibility_policy pair already exists)
+                raise JsonableError(_("Topic already muted"))
+
+            # The request is to just *change* the visibility policy of a topic
...skipping...
 from zerver.actions.user_topics import do_set_user_topic_visibility_policy
-from zerver.lib.exceptions import JsonableError
 from zerver.lib.request import REQ, has_request_variables
 from zerver.lib.response import json_success
 from zerver.lib.streams import (
@@ -34,16 +32,13 @@ def mute_topic(
         assert stream_id is not None
         (stream, sub) = access_stream_by_id(user_profile, stream_id)

-    try:
-        do_set_user_topic_visibility_policy(
-            user_profile,
-            stream,
-            topic_name,
-            visibility_policy=UserTopic.MUTED,
-            last_updated=date_muted,
-        )
-    except IntegrityError:
-        raise JsonableError(_("Topic already muted"))
+    do_set_user_topic_visibility_policy(
+        user_profile,
+        stream,
+        topic_name,
+        visibility_policy=UserTopic.MUTED,
+        last_updated=date_muted,
+    )


 def unmute_topic(
```

Reply:
Hmm, that's a better approach. -- Updated the PR